cobalt/examples/utils.cc

```diff
@@ -0,0 +1,40 @@
+// Copyright 2023 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/examples/utils.h"
+
+#include "starboard/common/log.h"
+#include "starboard/extension/time_zone.h"
+#include "starboard/file.h"
+#include "starboard/log.h"
+#include "starboard/system.h"
+
+bool StateChangingFunction(const char* path) {
+  SbLogFormatF("Going to delete: %s", path);
+  return SbFileDelete(path);
+}
+
+bool StateChangingFunctionWithSbExtensionDep(const char* time_zone_name) {
+  auto time_zone_extension = static_cast<const StarboardExtensionTimeZoneApi*>(
+      SbSystemGetExtension(kStarboardExtensionTimeZoneName));
+  if (time_zone_extension && time_zone_extension->version >= 1) {
+    // This log statement causes the unit test to fail because
```

Comment:
@jellefoks, I think this is an inherent disadvantage of using the link seam: it's difficult to limit the scope of the test double and prevent it from spilling over to places where you want/need to use the real implementation.

Here, the dependency - `SbSystemGetExtension` - is called both directly by the code-under-test, where we do want to stub/mock it, and by a low-level library used by the code-under-test, where we want to use the real implementation.

Let me know if you have any thoughts/ideas about this.

Reply:
The Stub can defer the call to the implementation for extensions with names that it's not handling.

Reply:
That woud fix this particular instance but it doesn't seem like a very general solution; the code-under-test and the low-level library involved here could easily have been using `SbSystemGetExtension` to access the same Starboard extension.

Reply:
Just following up to see if you have any more thoughts on this, @jellefoks?

Reply:
The link seam should be implemented such that a stub or mock implementation always has the ability to defer based on parameter inputs. For example, I can imagine that there is tremendous benefit for a mock be made for a networking call that adds delay, bandwidth, or reliability limitations, for which it would have to be able to defer after adding the desired additional functionality.

Reply:
I'm not following exactly how that would work. In the last example I gave, both callers are calling `SbSystemGetExtension(some_extension_name)`, so the proxy can't use the function argument - which is exactly `some_extension_name` in both calls - to decide whether it should defer to the stub/mock or to the real implementation.